File: search_compiler/compiler.py
from multiprocessing import Pool, cpu_count
from functools import partial
from itertools import chain
from timeit import default_timer as timer
import heapq

from .circuits import *

from . import solver as scsolver
from . import checkpoint, utils, heuristics, circuits, logging, gatesets

class Compiler():
    def compile(self, U, depth):
        raise NotImplementedError("Subclasses of Compiler are expected to implement the compile method.")
        return (U, None, None)

def evaluate_step(tup, U, error_func, error_jac, solver, I):
    step, depth = tup
    return (step, solver.solve_for_unitary(step, U, error_func), depth)

class SearchCompiler(Compiler):
    def __init__(self, threshold=1e-10, error_func=utils.matrix_distance_squared, error_jac=None, eval_func=None, heuristic=heuristics.astar, gateset=gatesets.Default(), solver=None, beams=-1, logger=None, verbosity=0):
        self.logger = logger
        self.verbosity = verbosity
        
        self.threshold = threshold
        self.error_func = error_func
        self.error_jac = error_jac

        if self.error_jac is None and error_func == utils.matrix_distance_squared:
            self.error_jac = utils.matrix_distance_squared_jac
        elif self.error_jac is None and error_func == utils.matrix_residuals:
            self.error_jac = utils.matrix_residuals_jac

        self.eval_func = eval_func

        if self.eval_func is None:
            if self.error_func == utils.matrix_residuals:
                self.eval_func = utils.matrix_distance_squared
            else:
                self.eval_func = self.error_func

        self.heuristic = heuristic
        self.gateset = gateset
        if solver is None:
            solver = scsolver.default_solver(gateset, 0, self.error_func, self.error_jac, self.logger)
            if type(solver) == scsolver.LeastSquares_Jac_Solver or type(solver) == scsolver.LeastSquares_Jac_SolverNative:
                # the default selector said we should switch to LeastSquares, so lets set the relevant values
                self.error_func = utils.matrix_residuals
                self.error_jac = utils.matrix_residuals_jac
                self.eval_func = utils.matrix_distance_squared

        self.solver = solver
        self.beams = int(beams)

    def compile(self, U, depth=None, statefile=None, logger=None):
        if logger is None:
            logger = self.logger
        if logger is None:
            logger = logging.Logger(stdout_enabled=True, verbosity=self.verbosity)

        startime = timer() # note, because all of this setup gets included in the total time, stopping and restarting the project may lead to time durations that are not representative of the runtime under normal conditions
        h = self.heuristic
        dits = int(np.round(np.log(np.shape(U)[0])/np.log(self.gateset.d)))

        if self.gateset.d**dits != np.shape(U)[0]:
            raise ValueError("The target matrix of size {} is not compatible with qudits of size {}.".format(np.shape(U)[0], self.d))

        I = circuits.IdentityStep(self.gateset.d)

        initial_layer = self.gateset.initial_layer(dits)
        search_layers = self.gateset.search_layers(dits)

        if len(search_layers) <= 0:
            logger.logprint("This gateset has no branching factor so only an initial optimization will be run.")
            root = initial_layer
            result = self.solver.solve_for_unitary(root, U, self.eval_func)
            return (result[0], root, result[1])


        logger.logprint("There are {} processors available to Pool.".format(cpu_count()))
        logger.logprint("The branching factor is {}.".format(len(search_layers)))
        beams = self.beams
        if self.beams < 1 and len(search_layers) > 0:
            beams = int(cpu_count() // len(search_layers))
        if beams < 1:
            beams = 1
        if beams > 1:
            logger.logprint("The beam factor is {}.".format(beams))
        pool = Pool(min(len(search_layers)*beams,cpu_count()))
        logger.logprint("Creating a pool of {} workers".format(pool._processes))

        recovered_state = checkpoint.recover(statefile)
        queue = []
        best_depth = 0
        best_value = 0
        best_pair  = 0
        tiebreaker = 0
        rectime = 0
        if recovered_state == None:
            root = ProductStep(initial_layer)
            result = self.solver.solve_for_unitary(root, U, self.error_func, self.error_jac)
            best_value = self.eval_func(U, result[0])
            best_pair = (root, result[1])
            logger.logprint("New best! {} at depth 0".format(best_value))
            if depth == 0:
                return best_pair

            queue = [(h(best_value, 0), 0, best_value, -1, result[1], root)]
            #         heuristic      depth  distance tiebreaker vector structure
            #             0            1      2         3         4        5
            checkpoint.save((queue, best_depth, best_value, best_pair, tiebreaker, timer()-startime), statefile)
        else:
            queue, best_depth, best_value, best_pair, tiebreaker, rectime = recovered_state
            logger.logprint("Recovered state with best result {} at depth {}".format(best_value, best_depth))

        while len(queue) > 0:
            if best_value < self.threshold:
                pool.close()
                pool.terminate()
                queue = []
                break
            popped = []
            for _ in range(0, beams):
                if len(queue) == 0:
                    break
                tup = heapq.heappop(queue)
                popped.append(tup)
                logger.logprint("Popped a node with score: {} at depth: {}".format((tup[2]), tup[1]), verbosity=2)

            then = timer()
            new_steps = [(current_tup[5].appending(search_layer), current_tup[1]) for search_layer in search_layers for current_tup in popped]

            for step, result, current_depth in pool.imap_unordered(partial(evaluate_step, U=U, error_func=self.error_func, error_jac=self.error_jac, solver=self.solver, I=I), new_steps):
                current_value = self.eval_func(U, result[0])
                if (current_value < best_value and (best_value >= self.threshold or current_depth + 1 <= best_depth)) or (current_value < self.threshold and current_depth + 1 < best_depth):
                    best_value = current_value
                    best_pair = (step, result[1])
                    best_depth = current_depth + 1
                    logger.logprint("New best! score: {} at depth: {}".format(best_value, current_depth + 1))
                if depth is None or current_depth + 1 < depth:
                    heapq.heappush(queue, (h(current_value, current_depth+1), current_depth+1, current_value, tiebreaker, result[1], step))
                    tiebreaker+=1
            logger.logprint("Layer completed after {} seconds".format(timer() - then), verbosity=2)
            checkpoint.save((queue, best_depth, best_value, best_pair, tiebreaker, rectime+(timer()-startime)), statefile)


        pool.close()
        pool.terminate()
        pool.join()
        logger.logprint("Finished compilation at depth {} with score {} after {} seconds.".format(best_depth, best_value, rectime+(timer()-startime)))
        return best_pair

File: qsrs/src/compiler.rs
use crate::circuits::*;
use crate::gatesets::Gateset;
use crate::heuristic::astar;
use crate::solvers::Solver;
use crate::utils::*;
use squaremat::SquareMatrix;

use std::cmp::Ordering;
use std::collections::BinaryHeap;
use std::convert::From;
use std::time::Instant;

pub trait Compiler {
    fn compile(&self, u: SquareMatrix, max_depth: Option<usize>) -> (SquareMatrix, Gate, Vec<f64>);
}

pub struct SearchCompiler<G>
where
    G: Gateset + Send + Sync,
{
    threshold: f64,
    gateset: G,
    beams: Option<usize>,
}

impl<G> SearchCompiler<G>
where
    G: Gateset + Send + Sync,
{
    pub fn new(threshold: f64, gateset: G, beams: Option<usize>) -> Self {
        Self {
            threshold,
            gateset,
            beams,
        }
    }
}

#[derive(PartialEq)]
struct QueueItem {
    heuristic: f64,
    depth: usize,
    distance: f64,
    tiebreaker: i64,
    params: Vec<f64>,
    circuit: GateProduct,
}

impl Eq for QueueItem {}

impl Ord for QueueItem {
    fn cmp(&self, other: &Self) -> Ordering {
        Ordering::reverse(
            self.heuristic.partial_cmp(&other.heuristic).unwrap_or(
                self.depth.partial_cmp(&other.depth).unwrap_or(
                    self.distance
                        .partial_cmp(&other.distance)
                        .unwrap_or(self.tiebreaker.cmp(&other.tiebreaker)),
                ),
            ),
        )
    }
}

impl PartialOrd for QueueItem {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl From<(f64, usize, f64, i64, Vec<f64>, GateProduct)> for QueueItem {
    fn from(item: (f64, usize, f64, i64, Vec<f64>, GateProduct)) -> Self {
        QueueItem {
            heuristic: item.0,
            depth: item.1,
            distance: item.2,
            tiebreaker: item.3,
            params: item.4,
            circuit: item.5,
        }
    }
}

impl<G> Compiler for SearchCompiler<G>
where
    G: Gateset + Send + Sync,
{
    fn compile(&self, u: SquareMatrix, max_depth: Option<usize>) -> (SquareMatrix, Gate, Vec<f64>) {
        let solv = crate::solvers::LeastSquaresJacSolver::new(1, 1e-6, 1e-10);
        let dits = (u.size as f64).log(self.gateset.d() as f64).round() as u8;
        assert_eq!(self.gateset.d().pow(dits as u32), u.size as usize);
        let initial_layer = self.gateset.initial_layer(dits);
        let search_layers = self.gateset.search_layers(dits);

        let cpus = num_cpus::get();
        println!("There are {} processors available in the Pool.", cpus);
        println!("The branching factor is {}.", search_layers.len());
        let beams = match self.beams {
            Some(b) => b,
            None => {
                if !search_layers.is_empty() {
                    cpus / search_layers.len()
                } else {
                    1
                }
            }
        };
        if beams > 1 {
            println!("The beam factor is {}.", beams);
        }
        let pool_size = cpus.min(search_layers.len() * beams);
        println!("Creating a pool of {} workers", pool_size);
        let start = Instant::now();
        // TODO checkpointing with serde
        let mut queue: BinaryHeap<QueueItem> = BinaryHeap::new();
        let mut best_depth = 0;
        let mut tiebreaker = -1;

        let root = GateProduct::new(vec![initial_layer]);
        let result = solv.solve_for_unitary(
            &root.clone().into(),
            &self.gateset.constant_gates(),
            &u,
            None,
        );
        let mut best_val = matrix_distance_squared(&u, &result.0);
        let mut best_pair = (result.0, root.clone().into(), result.1.clone());
        println!("New best! {} at depth 0", best_val);
        if let Some(0) = max_depth {
            return best_pair;
        };
        queue.push((astar(best_val, 0), 0, best_val, tiebreaker, result.1, root).into());

        while !queue.is_empty() {
            if best_val < self.threshold {
                queue.clear();
                break;
            }
            let mut popped = Vec::new();
            for _ in 0..beams {
                match queue.pop() {
                    Some(tup) => {
                        println!(
                            "Popped a node with score: {} at depth: {}",
                            tup.distance, tup.depth
                        );
                        popped.push(tup);
                    }
                    None => break,
                }
            }
            let then = Instant::now();

            let mut new_steps = Vec::with_capacity(popped.len() * search_layers.len());
            for current_tup in popped {
                for search_layer in search_layers.clone() {
                    new_steps.push((
                        current_tup.circuit.clone().append(search_layer.0),
                        current_tup.depth,
                    ))
                }
            }

            let results: Vec<(&GateProduct, (SquareMatrix, Vec<f64>), &usize)> = new_steps
                .iter()
                .map(|(step, depth)| {
                    let solv = crate::solvers::LeastSquaresJacSolver::new(1, 1e-6, 1e-10);
                    (
                        step,
                        solv.solve_for_unitary(
                            &step.clone().into(),
                            &self.gateset.constant_gates(),
                            &u,
                            None,
                        ),
                        depth,
                    )
                })
                .collect();
            for (step, result, current_depth) in results {
                let current_val = matrix_distance_squared(&u.clone(), &result.0);
                if (current_val < best_val
                    && (best_val >= self.threshold || current_depth < &best_depth))
                    || (current_val < self.threshold && current_depth + 1 < best_depth)
                {
                    best_val = current_val;
                    best_pair = (result.0, step.clone().into(), result.1.clone());
                    best_depth = current_depth + 1;
                    println!(
                        "New best! score: {:e} at depth: {}",
                        best_val,
                        current_depth + 1
                    );
                }
                if max_depth.is_none() || current_depth + 1 < max_depth.unwrap() {
                    queue.push(
                        (
                            astar(current_val, current_depth + 1),
                            current_depth + 1,
                            current_val,
                            tiebreaker,
                            result.1,
                            step.clone(),
                        )
                            .into(),
                    );
                    tiebreaker += 1;
                }
            }
            println!(
                "Layer completed after {} seconds",
                then.elapsed().as_millis() as f64 / 1000.0
            );
        }
        println!(
            "Finished compilation at depth {} with score {:e} after {} seconds.",
            best_depth,
            best_val,
            start.elapsed().as_millis() as f64 / 1000.0
        );
        best_pair
    }
}
